doc/design/design_cluster_job_runner.md

```diff
@@ -0,0 +1,126 @@
+1. The SQLFlow server is not High-Available, if an SQLFlow server instance failed, the jobs on this instance are failed.
+
+In this design, we propose to implement the **Cluster Job Runner** to solve the above problems.
+
+The cluster job runner would launch the job on Kubernetes or Yarn cluster instead of on the host. The SQLflow client can
+check the job status in a polling manner instead of a long live connection. We recommend using the cluster job runner in the production environment.
+
+## High-Level Design
+
+For the most submitter program, it can run as the local mode or distributed mode. For the local mode, SQLFlow would run
+the submitter program as a local process on the host; For the distributed model, the submitter program would submit a distributed Job to the cluster. For the two modes, the behavior of SQLFlow is different in both local job runner and cluster job runner:
+
+Job Runner| local mode | distributed mode
+-- | -- | --
+Local | launch a process on the host with blocking| submite a job to cluster with blocking
+Cluster| launch a Kubernetes Pod with no-blocing| submite a job to cluster with no-blocking
+
+The cluster job runner workflow is as follows:
+
+<img src="figures/cluster_job_runner.png">
+
+1. SQLFlow client sends the SQL statement via a gRPC call to the SQLFlow server. The local job goto **2** and the distributed
+job go to **3**.
+2. If the SQL statement implies a local job, e.g. local TensorFlow Job and local XGBoost job.
+    1. SQLFLow server would launch a Pod on Kubernetes cluster via Kubernetes API.
+    1. SQLFLow server can fetch the job tracker URL from the api-server.
+    1. The SQLFlow server returned the job tracker URL to the client.
+3. If the SQL statement implies a distributed job, e.g. distributed Tensorflow Job, distributed ALPS job, or other distributed
+machine learning job.
+    1. SQLFlow server would call the distributed AI framework API to launch a distributed Job on a cluster(Kubernetes/Yarn),
+    e.g. EDL API or ALPS API.
+    1. SQLFLow server can fetch the Job tracer URL from the distributed AI framework API.
+    1. SQLFlow server returns the job tracer URL to the client.
+4. The SQLFLow client would send the job tracer URL to the SQLFlow server to request the task status in a polling manner until
+the returned task status is **COMPLETED** or **FAILED**.
+
+## Proposal Details
+
+1. Add a new gRPC interface to check the job status.
+
+    ``` protobuf
+    service SQLFlow {
+      /**
+      if sqlflow_job_runner == "local":
+          sqlflow.Run(sql...)
+      elif sqlflow_job_runner == "cluster":
+          res = sqlflow.Run(sql...)
+          is_finised = False
+          while !is_finished:
+              is_finised = sqlflow.IsFinished(res.session.job_tracker_url)
+      **/
+      rpc IsFinished(Request) (stream TaskStatus) // add a new gRPC interface to request the job status
+      rpc Run (Request) returns (stream Response);
+    }
+
+    message Session {
```

Comment:
Can `Session` represent information of multiple SQLs?

Reply:
After putting more thoughts into this problem, I realize the problem boils down to the `job_tracker_url` only corresponding to a training job, while a `client.Run()` can create multiple jobs. With this been said, I don't think we can avoid maintaining states on the server end.

Reply:
For the local job runner, the SQL job is managed by the server-side, and maintain the states on the server.
For the cluster job runner, the SQL job is running in a Kubernetes Pod, the client checks the job by check the k8s Pod stats, so the server does not need to maintain the states.

And I think `job_tracker_url` or `token` is corresponding a SQL job, the SQL job may contain one or more SQL statements.